### PR TITLE
MKLdnn Integration Patch to improve issue #2986(call for cpu performance)

### PR DESCRIPTION
**Pull request by Darwin2011**
Hi, @piiswrong 

Mxnet's CPU performance is not as good as GPU's. 
please see #2986. @xmchen1987

This PR integrates MKLDNN into mxnet including all the major operations(conv, pool, lrn, bn and fc).
It turns out we can achieve 750+ fps for alexnet scoring and 280+ fps for Googlenet-v1 scoring with this patch in two-socket broadwell server and the backward also can benefit greatly from this patch. 

Please review the patches then give us feedbacks.
Thanks.

### CODE REVIEW

File: make/config.mk [hunk 1]
```diff
@@ -20,18 +20,21 @@
 # choice of compiler
 #--------------------
 
+
```
Comment (piiswrong) [inline make/config.mk]: revert changes in this file

File: mshadow.mk.tmp [hunk 2]
```diff
@@ -0,0 +1,104 @@
+#---------------------------------------------------------------------------------------
```
Comment (piiswrong) [inline mshadow.mk.tmp]: remove this file?

File: prepare_intel.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+mkdir -p ./external/mkl/
+cp prepare_mkl.sh ./external/mkl/
+./external/mkl/prepare_mkl.sh 0
+cp -f tpp.h external/mkl/mklml_lnx_2017.0.0.20160801/include/mkl.h
+cp -f config.mk.tmp make/config.mk
```
Comment (piiswrong) [inline prepare_intel.sh]: code these as conditions in makefile.
config.mk in root folder will override make/config.mk.
This approach won't work for existing isntallations

File: prepare_mkl.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
```
Comment (piiswrong) [inline prepare_mkl.sh]: hide this or merge it in prepare_intel.sh

File: setenv.sh [hunk 5]
```diff
@@ -0,0 +1 @@
+export LD_LIBRARY_PATH=${PWD}/external/mkl/mklml_lnx_2017.0.0.20160801/lib:$LD_LIBRARY_PATH
```
Comment (piiswrong) [inline setenv.sh]: write a readme and instruct the user to add this to their .bashrc.

Overall I think it's better to make mkl installation separate from mxnet.
say you have a install_mklml.sh that downloads, compiles, and installs mkl to PREFIX and ask the user to append LD_LIBRARY_PATH=PREFIX/lib to their bashrc.

Some users like to install mxnet into their python site-packages instead of using it inplace

File: src/operator/fully_connected-inl.h [hunk 6]
```diff
@@ -56,6 +56,9 @@ class FullyConnectedOp : public Operator {
                        const std::vector<TBlob> &aux_args) {
     using namespace mshadow;
     using namespace mshadow::expr;
+#ifdef PRINT_LAYER_TIME
```
Comment (piiswrong) [inline src/operator/fully_connected-inl.h]: we will have a profiler soon. This should be removed, but it's ok if you want to leave it here until the profiler comes in

File: src/operator/mkldnn_batch_norm-inl.h [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+#include <map>
+#include <vector>
+#include <string>
+#include <utility>
+#include "./operator_common.h"
+#include "./mshadow_op.h"
+#include "mkldnn_cppwrapper.h"
+#include "../mkl_memory-inl.h"
+#include "dmlc/timer.h"
+#include "memory"
```
Comment (piiswrong) [inline src/operator/mkldnn_batch_norm-inl.h]: what's this?

File: src/operator/pooling-inl.h [hunk 8]
```diff
@@ -199,8 +208,13 @@ class PoolingProp : public OperatorProperty {
         CHECK(param_.kernel[0] <= dshape[2] + 2 * param_.pad[0]
               && param_.kernel[1] <= dshape[3] + 2 * param_.pad[1])
             << "kernel size exceed input";
-        oshape[2] = 1 + (dshape[2] + 2 * param_.pad[0] - param_.kernel[0]) / param_.stride[0];
-        oshape[3] = 1 + (dshape[3] + 2 * param_.pad[1] - param_.kernel[1]) / param_.stride[1];
+        if (param_.pooling_convention == pool_enum::kValid) {
```
Comment (piiswrong) [inline src/operator/pooling-inl.h]: are these changes from another pr? do a rebase?

File: src/operator/mkldnn_pooling-inl.h [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+            1 + static_cast<int>(ceil(static_cast<float>(
+                    input_height + 2 * this->param_.pad[0] -
+                    this->param_.kernel[0]) / this->param_.stride[0]));
+        size_t full_model_output_width =
+            1 + static_cast<int>(ceil(static_cast<float>(
+                    input_width + 2 * this->param_.pad[0] -
+                    this->param_.kernel[0]) / this->param_.stride[0]));
+        if ((full_model_output_width == output_width) &&
+            (full_model_output_height == output_height)) {
+          support_mkldnn_ = true;
```
Comment (piiswrong) [inline src/operator/mkldnn_pooling-inl.h]: You can determine this in createoperaterex.
Instead of inheriting pooling, just return it if it's not supported by cudnn.

File: src/operator/pooling-inl.h [hunk 10]
```diff
@@ -25,13 +25,15 @@ namespace pool_enum {
...
 enum PoolingOpType {kMaxPooling, kAvgPooling, kSumPooling};
+enum PoolingOpPadConventionType {kValid, kFull};
 }  // namespace pool_enum
 
 struct PoolingParam : public dmlc::Parameter<PoolingParam> {
   TShape kernel;
   TShape stride;
   TShape pad;
   int pool_type;
+  int pooling_convention;
```
Comment (piiswrong) [inline src/operator/pooling-inl.h]: please rebase.

File: tests/python/unittest/test_operator.py [hunk 11]
```diff
@@ -1481,13 +1481,13 @@ def test_roipooling():
     test_maximum_minimum_scalar()
     test_abs()
     test_round_ceil_floor()
-    test_deconvolution()
+    #test_deconvolution()
```
Comment (piiswrong) [inline tests/python/unittest/test_operator.py]: revert these changes

File: tpp.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
+/*******************************************************************************
+* Copyright 1999-2016 Intel Corporation All Rights Reserved.
```
Comment (piiswrong) [inline tpp.h]: what's this? can we remove it? inserting this into a Apache software doesn't feel right

**Comment by piiswrong** [comment]
Made a few comments. Two general points:
1) move everthing mkl related into src/operators/mkl/. In the future we want to move it into plugins/
2) separate mkl installation from mxnet (you can leave a helper script in the root, but it should install mkl to a proper location)

**Comment by piiswrong** [comment]
also add author names into all files:
+/*!
 + * Copyright (c) 2015 by Contributors
 + * \file mkldnn_pooling-inl.h
 + * \brief
 + * \author Chen, Xiaoming
 +*/

File: .gitmodules [hunk 13]
```diff
@@ -1,6 +1,6 @@
 [submodule "mshadow"]
 	path = mshadow
-	url = https://github.com/dmlc/mshadow.git
+	url = https://github.com/xmchen1987/mshadow.git
```
Comment (Darwin2011) [inline .gitmodules]: Will remove it. We add this to reproduce easily performance number.

File: config.mk.tmp [hunk 14]
```diff
@@ -0,0 +1,138 @@
+#-------------------------------------------------------------------------------
```
Comment (Darwin2011) [inline config.mk.tmp]: Okay. I will create one separated repo for this build scripts.

File: include/mkldnn_cppwrapper.h [hunk 15]
```diff
@@ -0,0 +1,1153 @@
+#ifndef _MKL_DNN_CPPWRAPPER_H
```
Comment (Darwin2011) [inline include/mkldnn_cppwrapper.h]: Sure. NP.

File: config.mk.tmp (hunk 14 above)
Comment (piiswrong) [inline config.mk.tmp]: I didn't mean a separate repo. There should be a option in config.mk called USE_MKLDNN and all the other flags should be added in Makefile if you see USE_MKLDNN=1

File: include/mkldnn_cppwrapper.h [hunk 16]
```diff
@@ -0,0 +1,1153 @@
...
+#include <stddef.h>
+
+#include <mshadow/base.h>
+#include <dmlc/logging.h>
+
+#include "mkl_dnn_types.h"
+#include "mkl_dnn.h"
+#include "mkl_version.h"
+
+#define TEMPLATE_PREFIX template <typename Dtype> inline
```
Comment (Darwin2011) [inline include/mkldnn_cppwrapper.h]: Thanks.
Renaming is one better solution, I think.

File: src/operator/pooling-inl.h (hunk 10 above)
Comment (Darwin2011) [inline src/operator/pooling-inl.h]: Okay

File: tests/python/unittest/test_operator.py (hunk 11 above)
Comment (Darwin2011) [inline tests/python/unittest/test_operator.py]: Thanks, reverted.

File: config.mk.tmp (hunk 14 above)
Comment (Darwin2011) [inline config.mk.tmp]: ok, thanks.

File: make/config.mk (hunk 1 above)
Comment (Darwin2011) [inline make/config.mk]: reverted and add mkldnn option. done.

File: mshadow.mk.tmp (hunk 2 above)
Comment (Darwin2011) [inline mshadow.mk.tmp]: remove it.

File: prepare_intel.sh (hunk 3 above)
Comment (Darwin2011) [inline prepare_intel.sh]: I will add this scripts later on.

File: src/operator/fully_connected-inl.h (hunk 6 above)
Comment (Darwin2011) [inline src/operator/fully_connected-inl.h]: Thanks for your information. Remove it firstly.

File: src/operator/mkldnn/mkldnn_memory-inl.h [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+
+#ifndef MXNET_OPERATOR_MKLDNN_MKLDNN_MEMORY_INL_H_
+#define MXNET_OPERATOR_MKLDNN_MKLDNN_MEMORY_INL_H_
+
+#include <dmlc/logging.h>
+#include "./mkldnn_cppwrapper.h"
+namespace mxnet {
+#if MXNET_USE_MKLDNN == 1
+template <typename DType>
+struct MKLMemoryDescriptorBase {
```
Comment (piiswrong) [inline src/operator/mkldnn/mkldnn_memory-inl.h]: Does this hold actual input/output buffers or just descriptors? What's the memory used for?
If it's actual data, consider using mxnet's tempspace so that it can be shared to save memory.

This is probably a hard fix, we can do it in the next release

File: src/operator/mkldnn/mkldnn_pooling-inl.h [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        Tensor<cpu, 4, DType> out =
+            out_data[pool_enum::kOut].get<cpu, 4, DType>(s);
+        pooling_res_[dnnResourceSrc] =
+            reinterpret_cast<void *>(fwd_in_data_->get_converted_prv(data.dptr_, false));
+        pooling_res_[dnnResourceDst] =
+            reinterpret_cast<void *>(fwd_out_data_->get_converted_prv(out.dptr_, false));
+        CHECK_EQ(dnnExecute<DType>(poolingFwd_, pooling_res_), E_SUCCESS);
+        fwd_out_data_->get_output_ptr(data.dptr_);
+      } else {
+        PoolingOp<cpu, Reducer, DType>::Forward(ctx, in_data, req, out_data,
```
Comment (piiswrong) [inline src/operator/mkldnn/mkldnn_pooling-inl.h]: Can you determine support_mkldnn_ at operator creation time? If so please directly create a PoolingOp instead of forwarding inside mkldnn_pooling

File: src/operator/mkldnn/mkldnn_pooling-inl.h [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+            1 + static_cast<int>(ceil(static_cast<float>(
+                    input_height + 2 * this->param_.pad[0] -
+                    this->param_.kernel[0]) / this->param_.stride[0]));
+        size_t full_model_output_width =
+            1 + static_cast<int>(ceil(static_cast<float>(
+                    input_width + 2 * this->param_.pad[0] -
+                    this->param_.kernel[0]) / this->param_.stride[0]));
+        if ((full_model_output_width == output_width) &&
+            (full_model_output_height == output_height)) {
+          support_mkldnn_ = true;
```
Comment (piiswrong) [inline src/operator/mkldnn/mkldnn_pooling-inl.h]: This is basically equivalent to if(param.pooling_convention == kFull) in createoperatorex right?

File: src/operator/pooling-inl.h (hunk 8 above)
Comment (piiswrong) [inline src/operator/pooling-inl.h]: Ok so that PR is abandoned right?

**Comment by piiswrong** [comment]
Mostly looks good to me now. Please address these issues and we can merge:
1) Submit a PR to dmlc/mshadow https://github.com/dmlc/mshadow/pulls
2) Update this PR after 1) is merged.
3) Fix Pooling so it doesn't inherit PoolingOp, return PoolingOp in createoperatorEx if mkldnn doesn't support it
4) Fix test errors and lint

File: make/config.mk [hunk 20]
```diff
@@ -51,6 +51,12 @@ USE_CUDNN = 0
 # whether use cuda runtime compiling for writing kernels in native language (i.e. Python)
 USE_NVRTC = 0
 
+# whether use MKLDNN library
+USE_MKLDNN = 1
```
Comment (Darwin2011) [inline make/config.mk]: Thanks. Set MKLDNN to 0.

File: src/operator/pooling-inl.h (hunk 8 above)
Comment (Darwin2011) [inline src/operator/pooling-inl.h]: Yes. Thanks.

File: src/operator/mkldnn/mkldnn_memory-inl.h (hunk 17 above)
Comment (Darwin2011) [inline src/operator/mkldnn/mkldnn_memory-inl.h]: This holds actual buffers.
MKLDNN's layer output results stores as different layout and memory size(I guess more strict memory alignment), which cannot be saved directly into ndarray/tblob.
We also do some first trial to fix this. After this PR, hope we can discuss about it. Thanks.

File: src/operator/mkldnn/mkldnn_pooling-inl.h (hunk 19 above)
Comment (Darwin2011) [inline src/operator/mkldnn/mkldnn_pooling-inl.h]: No. 
As our experiments, MKLDNN only support rounds up pooling. 
But sometimes, rounds up pooling and rounds down pooling has same output size.
That piece of codes determines whether the output size is same as rounds up pooling.
I know it's confusing so update the pooling code. Hope it can be better.

**Comment by Darwin2011** [comment]
@piiswrong 
- Fix pooling layer done. 
- PR has proposed to mshadow.  Although openmp on can boost some performance, it seems that there's no way to avoid competition from multiple ops. Please see. 
https://github.com/dmlc/mshadow/pull/167

**Comment by piiswrong** [comment]
Are you sure mshadow pr has been proposed? I can't find it.
I mean the pr with changes related to this patch. as you can see here, it's not compiling because it cannot find your commit to mshadow: https://travis-ci.org/dmlc/mxnet/jobs/165290833

File: src/operator/mkldnn/mkldnn_memory-inl.h (hunk 17 above)
Comment (piiswrong) [inline src/operator/mkldnn/mkldnn_memory-inl.h]: after nnvm you can try to do this in passes. i.e. insert layout swap layer before and after all mkldnn layers

**Comment by Darwin2011** [comment]
@piiswrong 
We have PR the mshadow patch but it's not accepted yet.
So  I change the mshadow commit to current mshadow head commit to make it work now.
Thanks for your help.

**Comment by piiswrong** [comment]
build is still failing. Please make sure it compiles on a machine without mkldnn when USE_MKLDNN=0

**Comment by Darwin2011** [comment]
@piiswrong
 I have made first commit pass CI check but after README commit, CI tests fails in case of cpptests.
I am taking a look at the issue. And Could you give me some advice?
Thanks

**Comment by piiswrong** [comment]
cpp tests segfaults randomly on travis. Can't be reproduced locally so no fix yet. Try rerunning it

**Comment by Darwin2011** [comment]
Thanks. After rerunning, the issue disappears.

**Comment by howard0su** [comment]
this looks promising. any updates?
